File: backend/app/routers/results.py
from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.result import (
    ExerciseTypeCreate,
    ExerciseTypeResponse,
    ExerciseTypeUpdate,
    ResultEntryCreate,
    ResultEntryListResponse,
    ResultEntryResponse,
    ResultEntryUpdate,
    ResultTrendPoint,
    ResultTrendResponse,
)
from app.services import result as result_service

router = APIRouter(prefix="/api", tags=["results"])


# ---------------------------------------------------------------------------
# Exercise Types
# ---------------------------------------------------------------------------


@router.get("/exercise-types", response_model=list[ExerciseTypeResponse])
def list_exercise_types(db: Session = Depends(get_db)):
    return result_service.list_exercise_types(db)


@router.post("/exercise-types", response_model=ExerciseTypeResponse, status_code=201)
def create_exercise_type(data: ExerciseTypeCreate, db: Session = Depends(get_db)):
    return result_service.create_exercise_type(db, data)


@router.put("/exercise-types/{exercise_type_id}", response_model=ExerciseTypeResponse)
def update_exercise_type(
    exercise_type_id: str, data: ExerciseTypeUpdate, db: Session = Depends(get_db)
):
    return result_service.update_exercise_type(db, exercise_type_id, data)


@router.delete("/exercise-types/{exercise_type_id}", status_code=204)
def delete_exercise_type(exercise_type_id: str, db: Session = Depends(get_db)):
    result_service.delete_exercise_type(db, exercise_type_id)


# ---------------------------------------------------------------------------
# PR History & Trends (must be before parameterized /results/{entry_id})
# ---------------------------------------------------------------------------


@router.get("/results/prs/{exercise_type_id}", response_model=list[ResultEntryResponse])
def get_pr_history(exercise_type_id: str, db: Session = Depends(get_db)):
    return result_service.get_pr_history(db, exercise_type_id)


@router.get("/results/trends/{exercise_type_id}", response_model=ResultTrendResponse)
def get_result_trend(
    exercise_type_id: str,
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
):
    et, entries = result_service.get_result_trend(
        db, exercise_type_id, date_from=date_from, date_to=date_to
    )
    return ResultTrendResponse(
        exercise_type_id=et.id,
        exercise_name=et.name,
        result_unit=et.result_unit,
        data=[
            ResultTrendPoint(
                recorded_date=e.recorded_date, value=e.value, is_pr=e.is_pr, is_rx=e.is_rx
            )
            for e in entries
        ],
    )


# ---------------------------------------------------------------------------
# Result Entries
# ---------------------------------------------------------------------------


@router.get("/results", response_model=ResultEntryListResponse)
def list_results(
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    exercise_type_id: str | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
):
    items, total = result_service.list_result_entries(
        db,
        page=page,
        per_page=per_page,
        exercise_type_id=exercise_type_id,
        date_from=date_from,
        date_to=date_to,
    )
    return ResultEntryListResponse(items=items, total=total, page=page, per_page=per_page)


@router.get("/results/{entry_id}", response_model=ResultEntryResponse)
def get_result(entry_id: str, db: Session = Depends(get_db)):
    return result_service.get_result_entry(db, entry_id)


@router.post("/results", response_model=ResultEntryResponse, status_code=201)
def create_result(data: ResultEntryCreate, db: Session = Depends(get_db)):
    return result_service.create_result_entry(db, data)


@router.put("/results/{entry_id}", response_model=ResultEntryResponse)
def update_result(entry_id: str, data: ResultEntryUpdate, db: Session = Depends(get_db)):
    return result_service.update_result_entry(db, entry_id, data)


@router.delete("/results/{entry_id}", status_code=204)
def delete_result(entry_id: str, db: Session = Depends(get_db)):
    result_service.delete_result_entry(db, entry_id)

File: frontend/src/api/results.ts
import { api } from "./client";
import type {
  ExerciseType,
  ExerciseTypeCreate,
  ResultEntry,
  ResultEntryCreate,
  ResultEntryUpdate,
  ResultEntryListResponse,
  ResultTrendResponse,
} from "../types/result";

export function listExerciseTypes(params?: { tag?: string }): Promise<ExerciseType[]> {
  const search = new URLSearchParams();
  if (params?.tag) search.set("tag", params.tag);
  const qs = search.toString();
  return api.get<ExerciseType[]>(`/api/exercise-types${qs ? `?${qs}` : ""}`);
}

export function createExerciseType(data: ExerciseTypeCreate): Promise<ExerciseType> {
  return api.post<ExerciseType>("/api/exercise-types", data);
}

export function deleteExerciseType(id: string): Promise<void> {
  return api.delete(`/api/exercise-types/${encodeURIComponent(id)}`);
}

export function listResultEntries(params?: {
  page?: number;
  per_page?: number;
  exercise_type_id?: string;
  date_from?: string;
  date_to?: string;
}): Promise<ResultEntryListResponse> {
  const search = new URLSearchParams();
  if (params?.page) search.set("page", String(params.page));
  if (params?.per_page) search.set("per_page", String(params.per_page));
  if (params?.exercise_type_id) search.set("exercise_type_id", params.exercise_type_id);
  if (params?.date_from) search.set("date_from", params.date_from);
  if (params?.date_to) search.set("date_to", params.date_to);
  const qs = search.toString();
  return api.get<ResultEntryListResponse>(`/api/results${qs ? `?${qs}` : ""}`);
}

export function createResultEntry(data: ResultEntryCreate): Promise<ResultEntry> {
  return api.post<ResultEntry>("/api/results", data);
}

export function updateResultEntry(id: string, data: ResultEntryUpdate): Promise<ResultEntry> {
  return api.put<ResultEntry>(`/api/results/${encodeURIComponent(id)}`, data);
}

export function deleteResultEntry(id: string): Promise<void> {
  return api.delete(`/api/results/${encodeURIComponent(id)}`);
}

export function getResultTrend(
  exerciseTypeId: string,
  params?: { date_from?: string; date_to?: string },
): Promise<ResultTrendResponse> {
  const search = new URLSearchParams();
  if (params?.date_from) search.set("date_from", params.date_from);
  if (params?.date_to) search.set("date_to", params.date_to);
  const qs = search.toString();
  return api.get<ResultTrendResponse>(
    `/api/results/trends/${encodeURIComponent(exerciseTypeId)}${qs ? `?${qs}` : ""}`,
  );
}

export function getPRHistory(exerciseTypeId: string): Promise<ResultEntry[]> {
  return api.get<ResultEntry[]>(`/api/results/prs/${encodeURIComponent(exerciseTypeId)}`);
}
